src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+  // or a principal name) is a constant or variable.
+  // TODO(#673 aferr) it would be better to refactor the AST so that
+  // identifiers are a separate AST node with constant and non-constant
+  // children determined at parse time rather than here since
+  // this information is available at parse time.
+  static bool IsNameConstant(const std::string& id) {
+    if (id[0] == '"') {
+      return true;
+    } else {
+      // Numeric literals are constants and this is
+      // a numeric literal if all the characters are numbers.
+      for (const char& c : id) {
+        if (!isdigit(c)) return false;
+      }
+      return true;
+    }
+  }
+
+  static datalog::ArgumentType GetTypingOrFatal(
+      const TypeEnv& type_env, const std::string& argument_name) {
+    if (type_env.find(argument_name) == type_env.end()) {
+      LOG(FATAL) << "could not find typing for argument" << argument_name;
+    } else {
+      return type_env.find(argument_name)->second;
+    }
+  }
+
+  static datalog::RelationDeclaration GetDeclarationOrFatal(
+      const DeclEnv& decl_env, const std::string& predicate_name) {
+    if (decl_env.find(predicate_name) == decl_env.end()) {
+      LOG(FATAL) << "could not find declaration for relation" << predicate_name;
+    } else {
+      return decl_env.find(predicate_name)->second;
+    }
+  }
+
+  // This generates a set of says assertions that populate all the
+  // universes for all the speakers given the type environment. Essentially,
+  // it loops through the list of constant principals, and then all the
+  // constants of all types, and then generates a SaysAssertion that
+  // adds the constant to the universe for each speaker.
+  static std::vector<SaysAssertion> GetUniverseDefiningFacts(
+      const TypeEnv& type_env) {
+    std::vector<Principal> constant_principals;
+    for (const auto& typing : type_env) {
+      if (typing.second.kind() == datalog::ArgumentType::Kind::kPrincipal) {
+        constant_principals.push_back(Principal(typing.first));
+      }
+    }
+
+    std::vector<SaysAssertion> universe_defining_says_assertions;
+    for (const auto& prin : constant_principals) {
+      for (const auto& typing : type_env) {
+        universe_defining_says_assertions.push_back(SaysAssertion(
+            prin,  // speaker
+            {Assertion(Fact({}, MakeUniverseMembershipFact(typing.first,
+                                                           typing.second)))}));
+      }
+    }
+    return universe_defining_says_assertions;
+  }
+
+  // This visitor populates a mapping (flat_hash_map) of constant (literal)
+  // names to their types by looking at the relations to which the literals
+  // are applied and declarations of those relations (including the types
+  // of the parameters where they are applied). It takes as input
+  // a mapping from relation names to relation declarations.
+  class TypeEnvironmentGenerationVisitor
```

Comment:
Usually, in C++, it's a good rule to have only one meaty `class` definition per header. You can sprinkle in a few small supporting `struct`s and `class`es and `enum`s, but another non-trivial `class` should get its own header.

Reply:
I think this may now have 0 meaty classes in the header and so this is fixed now. I'll leave this one open to see if you agree with that.

Reply:
speculating that this is resolved and resolving...